Src/PCLMock/WhenContinuation.cs

```diff
@@ -346,15 +347,26 @@ internal override object Apply(object mockedObject, params object[] args)
         {
             Debug.Assert(args != null);
 
-            if (this.exception != null)
+            if (this.doAction != null)
             {
-                throw this.exception;
+                args = this.ValidateActionArgs("Do", this.doAction.GetMethodInfo().GetParameters().Select(x => x.ParameterType).ToArray(), args);
+
+                try
+                {
+                    this.doAction.DynamicInvoke(args);
+                }
+                catch (TargetInvocationException ex)
+                {
+                    // Unwrap the inner exception and preserve its stack trace.
+                    ExceptionDispatchInfo
+                        .Capture(ex.InnerException)
+                        .Throw();
+                }
```

Comment:
This is critical to ensuring the exception that is thrown by the `Do` invocation makes sense to the caller.